File: Development/demo_unitTest_with_mocks/python/test_user_service.py
import pytest
from unittest.mock import Mock
from user_service import UserService

@pytest.fixture
def mock_external_api():
    mock_api = Mock()
    mock_api.fetch_user.return_value = {"name": "John Doe"}
    return mock_api

def test_get_user_data(mock_external_api):
    service = UserService(external_api=mock_external_api)
    result = service.get_user_data(user_id=123)
    assert result == "User: John Doe"
    mock_external_api.fetch_user.assert_called_once_with(123)

def test_get_user_data_invalid_response():
    mock_api = Mock()
    mock_api.fetch_user.return_value = None

    service = UserService(external_api=mock_api)
    with pytest.raises(ValueError):
        service.get_user_data(user_id=123)

File: Development/demo_unitTest_with_mocks/nodejs/testUserService.js
const { expect } = require("chai");
const sinon = require("sinon");
const UserService = require("./userService");

describe("UserService Tests", () => {
    it("should return user data", async () => {
        const mockApi = {
            fetchUser: sinon.stub().resolves({ name: "John Doe" }),
        };

        const service = new UserService(mockApi);
        const result = await service.getUserData(123);

        expect(result).to.equal("User: John Doe");
        sinon.assert.calledOnceWithExactly(mockApi.fetchUser, 123);
    });

    it("should throw error for invalid response", async () => {
        const mockApi = {
            fetchUser: sinon.stub().resolves(null),
        };

        const service = new UserService(mockApi);

        try {
            await service.getUserData(123);
            throw new Error("Expected error not thrown");
        } catch (err) {
            expect(err.message).to.equal("Invalid response from external API");
        }
    });
});
